analysis/util/RatingMath.py

```diff
@@ -77,14 +90,59 @@ def set_exhaustive_log_parameters(a: float, c:float, d:float, p:float = 1.0) ->
     D = d
     P = p
 
-def get_handicap_adjustment(rating: float, handicap: int) -> float:
+
+def get_handicap_rank_difference(handicap: int, size: int, komi: float, rules: str) -> float:
     global HALF_STONE_HANDICAP
     global HALF_STONE_HANDICAP_FOR_ALL_RANKS
+    global COMPUTE_HANDICAP_VIA_KOMI_SMALL
+    global COMPUTE_HANDICAP_VIA_KOMI_19X19
+
+    if (COMPUTE_HANDICAP_VIA_KOMI_19X19 and size == 19) or (COMPUTE_HANDICAP_VIA_KOMI_SMALL and size != 19):
+        # Use a "even game komi" that allows for draws for computing ratings.
+        # It's okay to expect a draw.
+        stone_value = 12
+        if rules == "japanese" or rules == "korean":
+            # Territory scoring.
+            area_bonus = 0
+            komi_bonus = 0
+        else:
+            # Bonus for the area value of a stone in area scoring.
+            area_bonus = 1
+
+            # Chinese and AGA rules add extra komi when there's a handicap but
+            # don't store it in the 'komi' field.
+            if rules == "chinese":
+                komi_bonus = 1 * handicap
+            elif rules == "aga":
+                komi_bonus = 1 * (handicap - 1) if handicap else 0
+            else:
+                komi_bonus = 0
+
+        # Convert the handicap into an equivalent komi, and subtract it from
+        # the komi for an even game.
+        handicap_as_komi = ((stone_value / 2 + area_bonus) - (komi + komi_bonus) +
+                            ((stone_value + area_bonus) * handicap if handicap > 1 else 0))
+
+        # Convert back to a fractional handicap, using scaling factors of 3x
+        # and 6x for 9x9 and 13x13.
+        if size == 9:
+            return handicap_as_komi * 6 / stone_value
+        if size == 13:
+            return handicap_as_komi * 3 / stone_value
+        return handicap_as_komi / stone_value
+
     if HALF_STONE_HANDICAP_FOR_ALL_RANKS:
-        return rank_to_rating(rating_to_rank(rating) + (handicap - 0.5 if handicap > 0 else 0)) - rating
+        return handicap - 0.5 if handicap > 0 else 0
     if HALF_STONE_HANDICAP:
-        return rank_to_rating(rating_to_rank(rating) + (0.5 if handicap == 1 else handicap)) - rating
-    return rank_to_rating(rating_to_rank(rating) + handicap) - rating
+        return 0.5 if handicap == 1 else handicap
+    return handicap
+
+
+def get_handicap_adjustment(rating: float, handicap: int, size: int, komi: float, rules: str) -> float:
+    rank_difference = get_handicap_rank_difference(handicap, size, komi, rules)
+    effective_rank = min(39, max(0, rating_to_rank(rating) + rank_difference))
```

Comment:
I'm trying to ponder here, I *think* the reason why I didn't clamp the rank between 30k-9d for this was intentional (but not necessarily correct, that decision is certainly open to discussion).

My gut says that by clamping this we're going to introduce some unnecessary noise. There are legitimately terrible players (including purposefully terrible beginner bots), and there are some legitimately >9d "players" (think every strong KataGo bot). I call out the bots in specific because they have a whole lot of games and can actually move the needle when it comes to where a lot of players land with their ratings. While we clamp those for display to the users, that's more out of convention and UX than a good mathematical reason. I think for the purposes of doing good ratings I think it does a bit of harm to pretend like everyone is squished within those ranks.

Again, open to discussion if you have a good reason to clamp them.

Reply:
Main reasoning here was practical, to avoid a MathDomainError (can't remember if it was in `rank_to_rating` or in the subsequent ratings update calculation). I did NOT play around with the bounds of the clamping... maybe -10 (40k) to 45 would be sufficient to avoid the MathDomainError. Maybe these clamp values could be parameterized? Maybe we should depend on SkipLogic to preclude the need for this?

In terms of handling players with *true* ranks outside the UX, I'm not sure how reasonable results can be if the range (output) of `rating_to_rank` and the domain (input) of `rank_to_rating` don't match. But that's the kind of data I think I was seeing (after applying small board handicap).

Note that I saw the math coming up with effective handicap as high as 52, from an 8-stone 9x9 game early in OGS history (not sure there was only one....). (Maybe that error predated the `SkipLogic` change and wouldn't be necessary now...)

BTW, IMO,:
- For `rank_to_rating`/`rating_to_rank` calculations, we should choose a domain and range for `ranks` that match. Either we clamp for things outside of that, or don't rate games that handicaps put outside of that, or something.
- For UX, we can choose a smaller set of ranks to clamp to. So, the `rating_to_rank` might have a player down as 0 or -5, but the UX would show 25k (not 30k or 35k).

Not sure how that matches what we have right now... 

(Feels a bit like it would be worthwhile to consider more of these changes in isolation... maybe I should split up this PR more after all... not today though)

Reply:
> Main reasoning here was practical, to avoid a MathDomainError (can't remember if it was in `rank_to_rating` or in the subsequent ratings update calculation).

Don't have much time today, but my initial attempts to reproduce this (by commenting out the mitigation and the new SkipLogic code) haven't been successful...

Reply:
Converted back to a draft for now— since I can't reproduce the MathDomainError anymore it seems better to remove the clamping for now, and if we add it back it should be parameterized.

Reply:
Aha, found it, when I temporarily reverted the clamping, I inserted a bug in `get_handicap_adjustment`, which is why I couldn't reproduce.

The domain error is on tallygameanalytics:
```
Processing OGS data
   2,739,321 /   30,811,698 games processed.  192.3s remainingTraceback (most recent call last):
  File "/Users/dexonsmith/Repos/online-go/goratings/analysis/analyze_glicko2_one_game_at_a_time.py", line 99, in <module>
    tally.add_glicko2_analytics(analytics)
  File "/Users/dexonsmith/Repos/online-go/goratings/analysis/util/TallyGameAnalytics.py", line 102, in add_glicko2_analytics
    self.prediction_cost[size][speed][rank][handicap] += - math.log(result.expected_win_rate if black_won else 1 - result.expected_win_rate)
                                                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
ValueError: math domain error
```

Anyway, I still think the options can land without fixing the MathDomainError, and then we can decide whether to fix with clamping or to update the TalylGameAnalytics to handle an `expected_win_rate` of 0 (which I assume is what's causing the domain error). Sending a new PR soon.

Reply:
Makes sense, though honestly if we're hitting the domain error it's because our rating code is busted somewhere, so having it blow up isn't necessarily a bad thing (although having a more appropriate error would probably be better in an ideal world).